Store #[stable] attribute's `since` value in structured form

dtolnay opened the pull request:
Followup to https://github.com/rust-lang/rust/pull/116773#pullrequestreview-1680913901.

Prior to this PR, if you wrote an improper `since` version in a `stable` attribute, such as `#[stable(feature = "foo", since = "wat.0")]`, rustc would emit a diagnostic saying **_'since' must be a Rust version number, such as "1.31.0"_** and then throw out the whole `stable` attribute as if it weren't there. This strategy had 2 problems, both fixed in this PR:

1. If there was also a `#[deprecated]` attribute on the same item, rustc would want to enforce that the stabilization version is older than the deprecation version. This involved reparsing the `stable` attribute's `since` version, with a diagnostic **_invalid stability version found_** if it failed to parse. Of course this diagnostic was unreachable because an invalid `since` version would have already caused the `stable` attribute to be thrown out. This PR deletes that unreachable diagnostic.

2. By throwing out the `stable` attribute when `since` is invalid, you'd end up with a second diagnostic saying **_function has missing stability attribute_** even though your function is not missing a stability attribute. This PR preserves the `stable` attribute even when `since` cannot be parsed, avoiding the misleading second diagnostic.

Followups I plan to try next:

- Do the same for the `since` value of `#[deprecated]`.

- See whether it makes sense to also preserve `stable` and/or `unstable` attributes when they contain an invalid `feature`. What redundant/misleading diagnostics can this eliminate? What problems arise from not having a usable feature name for some API, in the situation that we're already failing compilation, so not concerned about anything that happens in downstream code?

dtolnay committed:
Store 'since' attribute as parsed Version

dtolnay committed:
Keep track of #[stable] attribute even if version cannot be parsed

dtolnay committed:
Update stability attribute sanity UI test to delete superfluous errors

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6633610681/job/18021636550) [(plain)](https://github.com/rust-lang/rust/commit/7b0e315ae6784c344d5c19a5bb83d96bc20f518e/checks/18021636550/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a7b10ec5-3bb2-46d4-9104-b6afdf88952d
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=sinceversion
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_a7b10ec5-3bb2-46d4-9104-b6afdf88952d
GITHUB_REF=refs/pull/117148/merge
GITHUB_REF_NAME=117148/merge
GITHUB_REF_PROTECTED=false
---
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0308]: mismatched types
   --> src/librustdoc/clean/types.rs:590:58
    |
590 |             StabilityLevel::Stable { since, .. } => Some(since),
    |                                                     ---- ^^^^^ expected `Symbol`, found `Since`
    |                                                     arguments to this enum variant are incorrect
    |
    |
help: the type constructed contains `Since` due to the type of the argument passed
    |
    |
590 |             StabilityLevel::Stable { since, .. } => Some(since),
    |                                                          |
    |                                                          |
    |                                                          this argument influences the type of `Some`
   --> /checkout/library/core/src/option.rs:571:5
    |
571 |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
    |     ^^^^
    |     ^^^^

error[E0308]: mismatched types
   --> src/librustdoc/clean/types.rs:597:58
    |
597 |             StabilityLevel::Stable { since, .. } => Some(since),
    |                                                     ---- ^^^^^ expected `Symbol`, found `Since`
    |                                                     arguments to this enum variant are incorrect
    |
    |
help: the type constructed contains `Since` due to the type of the argument passed
    |
    |
597 |             StabilityLevel::Stable { since, .. } => Some(since),
    |                                                          |
    |                                                          |
    |                                                          this argument influences the type of `Some`
   --> /checkout/library/core/src/option.rs:571:5
    |
571 |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
    |     ^^^^
    |     ^^^^

error[E0308]: mismatched types
   --> src/librustdoc/html/render/mod.rs:932:31
    |
932 |             if Some(since) != containing_const_ver =>
    |                               ^^^^^^^^^^^^^^^^^^^^ expected `Option<Since>`, found `Option<Symbol>`
    = note: expected enum `std::option::Option<Since>`
               found enum `std::option::Option<rustc_span::Symbol>`

error[E0277]: `Since` doesn't implement `std::fmt::Display`
error[E0277]: `Since` doesn't implement `std::fmt::Display`
   --> src/librustdoc/html/render/mod.rs:934:40
    |
934 |             Some((format!("const since {since}"), format!("const: {since}")))
    |                   |                    |
    |                   |                    |
    |                   |                    `Since` cannot be formatted with the default formatter
    |
   ::: /checkout/library/alloc/src/macros.rs:123:1
    |
123 | macro_rules! format {
---

error[E0277]: `Since` doesn't implement `std::fmt::Display`
   --> src/librustdoc/html/render/mod.rs:934:67
    |
934 |             Some((format!("const since {since}"), format!("const: {since}")))
    |                                                   |               |
    |                                                   |               |
    |                                                   |               `Since` cannot be formatted with the default formatter
    |
   ::: /checkout/library/alloc/src/macros.rs:123:1
    |
123 | macro_rules! format {
```

</details>

dtolnay committed:
Handle structured stable attribute 'since' version in rustdoc

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6634171989/job/18023184660) [(plain)](https://github.com/rust-lang/rust/commit/6933a671d32f6f8077404e6d558c7ba60bdbbfb1/checks/18023184660/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_7f4376b7-4fa4-4dfa-8900-a2bba459839c
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=sinceversion
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_7f4376b7-4fa4-4dfa-8900-a2bba459839c
GITHUB_REF=refs/pull/117148/merge
GITHUB_REF_NAME=117148/merge
GITHUB_REF_PROTECTED=false
---
   --> src/tools/clippy/clippy_utils/src/qualify_min_const_fn.rs:376:22
    |
375 |                   let short_version = since
    |  _____________________________________-
376 | |                     .as_str()
    | |                     -^^^^^^ method not found in `Since`
    | 

For more information about this error, try `rustc --explain E0599`.
error: could not compile `clippy_utils` (lib) due to previous error
```

</details>

dtolnay committed:
Handle structured stable attribute 'since' version in clippy

dtolnay committed:
CFG_RELEASE is mandatory, no need for option_env

The same file already contains another env!("CFG_RELEASE") on line 632,
so it's impossible for this crate to compile without CFG_RELEASE set.

dtolnay committed:
Expose a non-Symbol way to access current rustc version string

dtolnay commented on compiler/rustc_attr/src/builtin.rs:
```diff
@@ -158,6 +158,8 @@ pub enum Since {
     /// Stabilized in the upcoming version, whatever number that is.
     Current,
+    /// Failed to parse a stabilization version.
+    Err,
```
I based this on the `ExprKind::Err` from rustc_ast, which does not hold `ErrorGuaranteed`.

https://github.com/rust-lang/rust/blob/964ff0125b0933699d5e1e796fcbdd72ef7246be/compiler/rustc_ast/src/ast.rs#L1494-L1495

dtolnay commented on compiler/rustc_attr/src/builtin.rs:
```diff
@@ -25,8 +25,7 @@ use crate::session_diagnostics::{self, IncorrectReprFormatGenericCause};
 pub fn rust_version_symbol() -> Symbol {
-    let version = option_env!("CFG_RELEASE").unwrap_or("<current>");
-    Symbol::intern(&version)
+    Symbol::intern(env!("CFG_RELEASE"))
```
Yes, I'll take care of that.

dtolnay replied:
#117188

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 1a9ea1f1a591915ef58931496d998871b8dd5524 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 1a9ea1f1a591915ef58931496d998871b8dd5524 -->
<!-- homu: {"type":"Approved","sha":"1a9ea1f1a591915ef58931496d998871b8dd5524","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1a9ea1f1a591915ef58931496d998871b8dd5524 with merge 104ac7bb6a17aee95fed2f70f2100e9236c11710...
<!-- homu: {"type":"BuildStarted","head_sha":"1a9ea1f1a591915ef58931496d998871b8dd5524","merge_sha":"104ac7bb6a17aee95fed2f70f2100e9236c11710"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6650574441/job/18074328923)
Approved by: cjgillot
Pushing 104ac7bb6a17aee95fed2f70f2100e9236c11710 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6650574441/job/18074328923"},"merge_sha":"104ac7bb6a17aee95fed2f70f2100e9236c11710"} -->

rust-timer commented:
Finished benchmarking commit (104ac7bb6a17aee95fed2f70f2100e9236c11710): [comparison URL](https://perf.rust-lang.org/compare.html?start=ccb160d3432dff025ab62696a82e576bf08aa006&end=104ac7bb6a17aee95fed2f70f2100e9236c11710&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ccb160d3432dff025ab62696a82e576bf08aa006&end=104ac7bb6a17aee95fed2f70f2100e9236c11710&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ccb160d3432dff025ab62696a82e576bf08aa006&end=104ac7bb6a17aee95fed2f70f2100e9236c11710&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.9%, -0.7%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ccb160d3432dff025ab62696a82e576bf08aa006&end=104ac7bb6a17aee95fed2f70f2100e9236c11710&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range         | count |
|:----------------------------------:|:-----:|:-------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -             | 0     |
| Regressions ❌ <br /> (secondary)  | 10.2% | [9.6%, 10.8%] | 2     |
| Improvements ✅ <br /> (primary)   | -     | -             | 0     |
| Improvements ✅ <br /> (secondary) | -     | -             | 0     |
| All ❌✅ (primary)                 | -     | -             | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ccb160d3432dff025ab62696a82e576bf08aa006&end=104ac7bb6a17aee95fed2f70f2100e9236c11710&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data
**Artifact size**: 304.45 MiB -> 304.42 MiB (-0.01%)
